### PR TITLE
gh-101283: Try to load the fallback cmd.exe by an absolute path

### PR DESCRIPTION
**Pull request by arhadthedev**
* Issue: gh-101283

### CODE REVIEW

**Comment by arhadthedev** [comment]
> PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'D:\\a\\cpython\\cpython\\build\\test_python_2036�\\test_python_worker_5676�'

Looks like a fluke; restarting.

File: Lib/subprocess.py
Comment (eryksun) [inline Lib/subprocess.py]: This shouldn't assume the default location of the Windows directory, i.e. the base/root directory of the OS on the system drive. It can be installed elsewhere. We should probably implement `GetSystemDirectoryW()`, `GetSystemWindowsDirectoryW()`, and `GetwindowsDirectoryW()`[^1] in the `_winapi` module. Barring that, we should make a best effort to use the "SystemRoot" and "SystemDrive" environment variables if they're defined and not *empty strings*.

```suggestion
                system_drive = os.environ.get('SystemDrive') or 'C:'
                system_root = os.environ.get('SystemRoot') or os.path.join(
                    system_drive, 'Windows')
                comspec = os.environ.get('ComSpec') or os.path.join(
                    system_root, 'System32', 'cmd.exe')
```

[^1]: On a terminal server (i.e. a server that supports remote desktop sessions), the path that's returned by `GetwindowsDirectoryW()` will be "%UserProfile%\\Windows" instead of the default location "%SystemDrive%\\Windows" if the application executable lacks the flag `IMAGE_DLLCHARACTER­ISTICS_TERMINAL_SERVER_AWARE`. This should never be the case for the normal "python[w].exe" binaries. It could be the case if Python is embedded. However, it's extremely unlikely nowadays that any PE image lacks this flag. VC++ has set it by default for many years.

File: Lib/subprocess.py
Comment (eryksun) [inline Lib/subprocess.py]:
```suggestion
                    # or the system environment variables are broken.
```

File: Lib/subprocess.py
Comment (eryksun) [inline Lib/subprocess.py]: Starting a process is already expensive on Windows. We should avoid the cost of `os.stat()` in the common case. We always use the `ComSpec` environment variable if it's defined and not an empty string. The fallback to "%SystemRoot%\\System32\\cmd.exe" or "%SystemDrive%\\Windows\\System32\\cmd.exe" should be checked before falling back on "cmd.exe". (Alternatively, the `_winapi.CreateProcess()` call could be retried if it fails with `FileNotFoundError` with `shell=True`, but that's I think needlessly complicated for this fallback case. I just don't want to slow down the common, expected path.)

Also, if `comspec` is a fully qualified path and `executable` isn't set, we can set `executable` to let `CreateProcessW()` skip its internal `SearchPathW()` call.

```suggestion
                comspec = os.environ.get('ComSpec')
                if not comspec:
                    system_drive = os.environ.get('SystemDrive') or 'C:'
                    system_root = os.environ.get('SystemRoot') or os.path.join(
                        system_drive, 'Windows')
                    comspec = os.path.join(system_root, 'System32', 'cmd.exe')
                    if not os.path.isfile(comspec):
                        # cmd.exe is missing, or the system environment
                        # variables are broken, or they're undefined and the
                        # system is installed into a non-standard location.
                        # This is highly unlikely, and we cannot help here.
                        comspec = 'cmd.exe'
                if not executable and os.path.isabs(comspec):
                    executable = comspec
```

**Comment by arhadthedev** [comment]
@eryksun Thank you for improvements! Since none of my code is left, I've reattributed the news entry to you.

File: Lib/subprocess.py
Comment (eryksun) [inline Lib/subprocess.py]:
```suggestion
                # gh-101283: without a fully-qualified path, before Windows
                # checks the system directories, it first looks in the
                # application directory, and also the current directory if
                # NeedCurrentDirectoryForExePathW(ExeName) is true, so try
                # to avoid executing unqualified "cmd.exe".
```

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (zooba) [inline Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: This much detail belongs in the `subprocess` documentation, not here. This entry can be simply ```Uses a safer approach to find ``cmd.exe`` in :mod:`subprocess` when launching with ``shell=True``.```

Comment (zooba) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: Also the `C\Windows` mention is missing a colon.

File: Lib/subprocess.py
Comment (zooba) [inline Lib/subprocess.py]: I think this is overcomplicated. Just fall back on `SystemRoot` and if that's empty then just raise `FileNotFoundError`.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (zooba) [inline Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: @eryksun Are you okay with this? I think it's fine to leave it out - there doesn't seem to be more than the usual amount of reviewing here, and I know you don't normally seek credit like this ;)

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (zooba) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: Also, I think it should be documented as using "other mechanisms to locate the system `cmd.exe`, even if the process environment has been changed" so that we don't accidentally promise to rely on other environment variables.

File: Lib/subprocess.py
Comment (zooba) [inline Lib/subprocess.py]: If the environment is so broken, we're best off just failing. And that's going to happen immediately anyway, so we can just let the path go through without checking.

File: Lib/subprocess.py
Comment (zooba) [inline Lib/subprocess.py]: Might be worth adding to the documentation that specifying both `executable` and `shell=True` will launch `executable` and assume that it works like `cmd.exe`. This may be useful to use a non-default shell, though it would still be better to just launch such a shell directly (without using `shell=True`).

File: Lib/subprocess.py
Comment (gpshead) [inline Lib/subprocess.py]: nit: are environment variables case insensitive on Windows?  is that true within os.environ?  This is not mentioned in https://docs.python.org/3/library/os.html#os.environ.

File: Lib/subprocess.py
Comment (gpshead) [inline Lib/subprocess.py]: Should a RuntimeWarning be raised or logging.warn call made in this case as an FYI?

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (gpshead) [inline Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: It might be good to clarify this as "... if system environment variables from :data:`os.environ` are ..." to disambiguate from anyone thinking that the Popen `env=` parameter is used.

File: Lib/subprocess.py
Comment (eryksun) [reply Lib/subprocess.py]: Yes, environment variable names are case insensitive and case preserving when accessed via the API functions `GetEnvironmentVariableW()` and `SetEnvironmentVariableW()`. 

The `os.environ` mapping implements case-insensitive access by storing the keys as uppercase strings. It sets environment variables via `os.putenv()`, which ultimately calls `SetEnvironmentVariableW()`. The latter preserves the original name when updating the value of an existing variable.

A small issue, which is mostly a matter of aesthetics, concerns using `os.environ` as the `env` parameter of `subprocess.Popen`. The child process gets an environment block with uppercase variable names. We could implement `os.environ` more cleverly to get a case insensitive, case preserving dict.

File: Lib/subprocess.py
Comment (eryksun) [reply Lib/subprocess.py]: That's a good idea. How about this?

```python
                        comspec = 'cmd.exe'
                        warnings.warn(f'spawning "{comspec}" using a '
                                      'relative file path',
                                      RuntimeWarning)
```

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (eryksun) [inline Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst:6]:
```suggestion
On Windows, :meth:`subprocess.Popen.__init__` with ``shell=True`` 
now tries to use a full-qualified path for "cmd.exe" when ``ComSpec``
isn't defined. It looks for "cmd.exe" in its expected location 
relative to ``SystemRoot`` or ``SystemDrive`` in ``os.environ`` or, 
if neither is defined, relative to ``"C:\\Windows"``. This helps to 
protect against the system executing a malicious "cmd.exe" program
in the application directory or current working directory.  
Patch by Eryk Sun.
```

**Review by zooba — COMMENTED** [review]
Definitely needs the documentation change. I can be convinced that we need the more complicated fallback algorithm, but I think it's fine to make it simpler, provided we don't precisely specify it in the docs (so that we can change it in the future if needed).

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (zooba) [inline Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: This is the kind of too-much-detail I was concerned about in my review comments. Let's keep the NEWS entry simple and update the docs, but make sure we retain the ability to use an even more reliable algorithm in the future.

File: Lib/subprocess.py
Comment (eryksun) [reply Lib/subprocess.py]: This is a long-standing platform inconsistency. On POSIX, combining `executable` with `shell=True` replaces the shell path in `args`. It should have been the same on Windows. Instead we allow an arbitrary `exectuable` to be mismatched with a command-line based on `ComSpec`. I wouldn't want that behavior to be documented. I'd prefer to eliminate the inconsistency in 3.12 and document consistent, sensible behavior.

If the caller didn't pass `executable`, I'm suggesting to use it in order to speed up the `CreateProcessW()` call a bit. When `lpApplicationName` (i.e. `executable`) is `NULL`, `CreateProcessW()` has to parse the command name out of the command line and call `SearchPathW()`. That step gets skipped if `lpApplicationName` isn't `NULL`, in which case the path name of the executable gets resolved against the current working directory.

File: Lib/subprocess.py
Comment (gpshead) [reply Lib/subprocess.py]: Agreed for the `main` branch. I think it is worth keeping as a fallback in the release branch backports to avoid breaking someones working code.

I'm not the right person to understand if such a non \Windows\System32 based environment without the environment variables set actually ever exists in the real world.

File: Lib/subprocess.py
Comment (gpshead) [reply Lib/subprocess.py]: if we keep the fallback at all rather than just make it an error, something like that.  probably with a stacklevel= set.

related to my other comment, i wouldn't want to add a new warning in a security backport either.

File: Lib/subprocess.py
Comment (zooba) [reply Lib/subprocess.py]: It's basically already documented today that `executable` will override the launched process but not the command line arguments (I've actually used that deliberately for testing commands that inspect their own argv[0]), so all I'm proposing is clarifying that specifying it with `shell=True` will still do that and it may not be a good idea.

I sure don't want to start messing with the command line based on `executable`. It's so useful

File: Lib/subprocess.py
Comment (zooba) [reply Lib/subprocess.py]: Unfortunately, it's really easy to spawn a subprocess and specify a partial environment that doesn't inherit anything. Typically the child process won't even be able to start under these circumstances, but depending on how someone fixes it can affect how likely this is (e.g. add variables until it works? Might not add `COMSPEC`).

File: Lib/subprocess.py
Comment (eryksun) [reply Lib/subprocess.py]: > all I'm proposing is clarifying that specifying it with `shell=True` will still do that and it may not be a good idea.

I'd rather fix it in 3.12 and document the correct behavior. On POSIX with `shell=True`, using `executable` replaces the shell in the command line, which is the sensible thing to do. On Windows, that would entail using `executable` instead of the value of the `ComSpec` environment variable, but that was never implemented.

File: Lib/subprocess.py
Comment (eryksun) [reply Lib/subprocess.py]: This would significantly change the behavior. Currently we let `CreateProcessW()` search for "cmd.exe" if `ComSpec` isn't defined. The API doesn't use the `SystemRoot` or `SystemDrive` environment variables in `RtlGetExePath()`, the function that builds the executable search path. It builds the search path as follows:

* application directory, i.e. the dynamic environment variable `__APPDIR__` 
* current directory, i.e. ("."), if `NeedCurrentDirectoryForExePathW(name)` is true
* system directory, i.e. `GetSystemDirectoryW()`; e.g. "C:\\Windows\\System32"
* legacy WIN16 system directory (no API query); e.g. "C:\\Windows\\System"
* Windows directory, i.e. `GetSystemWindowsDirectoryW()`; e.g. "C:\\Windows"
* `PATH`

I'd prefer to fall back on using the system directory from `GetSystemDirectoryW()`. If we use that, then I'm comfortable with letting `_winapi.CreateProcess()` fail if "cmd.exe" isn't found there. I'm less comfortable with giving up if the `SystemRoot` or `SystemDrive` environment variable is wrong. I don't think that's the same as `ComSpec` being wrong, which is a shell path that we've been explicitly told to use. If `ComSpec` is wrong, failing is the best response.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (eryksun) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: It seemed important to Oleg to credit me, so I left it in place in the suggested revision. It wasn't my goal in making suggestions to take credit. I don't expect them to be accepted verbatim, or even accepted at all. They can be resolved as rejected or with a completely new take on the problem. :shrug:

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (arhadthedev) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: I've simplified the wording as @zooba advised so it isn't applicable anymore.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (arhadthedev) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: How should we document the change in `docs.python.org`? I suspect it cannot go under `Changed in version 3.x` because it will span through ≥3.7.17, ≥3.8.17, ≥3.9.17, ≥3.10.9, and ≥3.11.2.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (arhadthedev) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: @zooba Eryk greatly expanded my original fixed-path-only approach effectively replacing it with the dynamic one. I see it as a different implementation of the same idea so that's why I've attributed him.

On a second thought, though, I agree that totally omitting myself as an author of the original patch was not the right way, so I've added myself back.

File: Lib/subprocess.py
Comment (zooba) [reply Lib/subprocess.py]: Actually yeah, that makes sense. Somehow I blanked out that we add the first argument in ourselves anyway.

File: Lib/subprocess.py
Comment (zooba) [reply Lib/subprocess.py]: I'm also okay with failing without `ComSpec` and not using `SystemRoot`.

One other mid-ground that might save us adding a native method right away could be to read `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Session Manager\Environment@ComSpec` directly, which at least requires an administrator to modify, rather than just the parent process.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (zooba) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: Backporting this kind of change requires manual effort, unfortunately. List all the changed versions in `main`, then as we backport we'll drop any versions that are newer than the target branch.

File: Lib/subprocess.py
Comment (eryksun) [reply Lib/subprocess.py]: FYI, for `os.system(command)`, if `ComSpec` is defined and not empty, the `_wsystem()` call  tries `_wspawnv(_P_WAIT, comspec, {comspec, slash_c, command, NULL})`. The latter checks for the existence of `comspec` relative to the current working directory. If the filename has no extension, it successively checks `comspec` plus ".com", ".exe", ".bat", and ".cmd". It executes the first name that exists. If the `_wspawnv()` call fails, `_wsystem()` falls back on `_wspawnvp(_P_WAIT, L"cmd.exe", {L"cmd.exe", slash_c, command, NULL})`. This successively tries `_wspawnv()` for the current directory and each of the `PATH` directories. The C runtime thus has an unsafe fallback. That's yet another reason to implement `os.system()` with `subprocess.Popen`, as was already implemented for `os.popen()`.

Comment (eryksun) [reply Lib/subprocess.py]: The "ComSpec" value in the session manager's "Environment" registry key is a good default[^1]. That way we'll use the system shell that was configured by an administrator (e.g. maybe it's [TCC/LE](https://jpsoft.com/products/tcc-le.html)), rather than defaulting to "cmd.exe".

The default installed value of `ComSpec` is the `REG_EXPAND_SZ` value "%SystemRoot%\\system32\\cmd.exe". We can expand the system `ComSpec` value with `winreg.ExpandEnvironmentStrings()`. This API function doesn't have any special handling for known system variables such as `SystemRoot`, but it does have the advantage that it doesn't expand an environment variable if the value is an empty string. `ntpath.expandvars()`, in contrast, will expand an environment variable to an empty string.

If we're implementing strict behavior, we could fail immediately if the default shell path isn't a fully-qualified path with a drive and root. But leave it up to `CreateProcessW()` to check for existence.

[^1]: Though I'd rather we supported [`CreateEnvironmentBlock()`](https://learn.microsoft.com/en-us/windows/win32/api/userenv/nf-userenv-createenvironmentblock) in `os`. For example, `os.create_environ(system_only=False, inherit=False)`. The result would be an `os._Environ` mapping with keys that are case insensitive. If [issue 7310](https://github.com/python/cpython/issues/73010) is resolved, the keys will also be case preserving.

Comment (eryksun) [reply Lib/subprocess.py]: Here's an example that moves the work into a method. This example disallows using a relative path for the shell, unless it's from the `ComSpec` environment variable. If it has to fall back on "%SystemRoot%\\System32\\cmd.exe", it will fail if `SystemRoot` is missing or empty.

```python
def _isabspath(self, s):
    # ntpath.isabs() returns true for a rooted path such as "\System32".
    # A rooted path depends on the drive of the current working
    # directory, or, if it's the target path of a symlink, on the drive
    # of the opened path of the symlink. This method implements the
    # correct behavior. 
    s = s[:3].replace('/', '\\')
    return s[:2] == '\\\\' or (s[1:] == ':\\' and s[:1] != '\\')

def _get_shell_path(self):
    # gh-101283: avoid using a relative path for the shell, unless the
    # the current ComSpec environment variable is a relative path.
    # Fall back on the system's configured ComSpec value, or else on
    # "%SystemRoot%\System32\cmd.exe".
    comspec = os.environ.get('ComSpec')
    if comspec:
        return comspec
    import winreg
    try:
        with winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE,
                r'System\CurrentControlSet\Control'
                r'\Session Manager\Environment') as hkey:
            comspec, vt = winreg.QueryValueEx(hkey, 'ComSpec')
            if not comspec or vt not in (
                    winreg.REG_EXPAND_SZ, winreg.REG_SZ):
                raise ValueError
            if vt == winreg.REG_EXPAND_SZ:
                comspec = winreg.ExpandEnvironmentStrings(comspec)
    except (OSError, ValueError):
        comspec = winreg.ExpandEnvironmentStrings(
                    r'%SystemRoot%\System32\cmd.exe')
    if not self._isabspath(comspec):
        raise ValueError('default shell path is not absolute: '
                         f'{comspec!r}')
    return comspec
```

Usage:

```python
                shell_path = self._get_shell_path()
                if self._isabspath(shell_path):
                    executable = shell_path
                args = f'"{shell_path}" /c "{args}"'
```

Note that `shell_path` in the command line has to be quoted, in case the path contains spaces. That's true even if we use `executable` because the shell has to parse its command line.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (arhadthedev) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: How should I mention multiple versions?

- `.. versionchanged:: 3.7.17, 3.8.17, 3.9.17, 3.10.9, and 3.11.2`break rendering:

   > 

- comma-less `.. versionchanged:: 3.7.17 3.8.17 3.9.17 3.10.9 3.11.2` doesn't work too:

   > Безымянный

For reference, that's a random single-version entry:

> image

Comment (arhadthedev) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: One-per-line versions don't help either:

```
   .. versionchanged:: 3.7.17
   .. versionchanged:: 3.8.17
   .. versionchanged:: 3.9.17
   .. versionchanged:: 3.10.9
   .. versionchanged:: 3.11.2

      Changed Windows shell search [...]
```

>

Comment (gpshead) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: Just mention the current main branch version (3.12), it is normal to manually edit these documentation tags as appropriate in each of the bugfix backports.

**Comment by arhadthedev** [comment]
Merging the main to restart potentially flaky `test_concurrent_futures`.

File: Lib/subprocess.py
Comment (arhadthedev) [reply Lib/subprocess.py:1498]: @gpshead Thank you for the advice; applied.

File: Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst
Comment (arhadthedev) [reply Misc/NEWS.d/next/Security/2023-01-24-16-12-00.gh-issue-101283.9tqu39.rst]: @gpshead Thank you for the advice; applied.

File: Lib/subprocess.py
Comment (arhadthedev) [reply Lib/subprocess.py]: Done. BTW will the new note below be good?

```rst
   .. versionchanged:: 3.11.2

      Changed Windows shell search order for ``shell=True``. The current
      directory and ``%PATH%`` are no longer used when ``%COMSPEC%`` is not set;
      ``%SystemRoot%\System32\cmd.exe`` is now the only last resort instead.
      As a result, dropping a malicious program named ``cmd.exe`` into a current
      directory no longe works.
```

cc @eryksun

File: Lib/subprocess.py
Comment (zooba) [inline Lib/subprocess.py]:
```suggestion
                    system_root = os.environ.get('SystemRoot', '')
```

Just to avoid a TypeError on the next line if the variable is not set.

File: Lib/subprocess.py
Comment (zooba) [inline Lib/subprocess.py]:
```suggestion
                    if not os.path.isabs(comspec) or not os.path.isfile(comspec):
```

Avoid testing a relative path at all and just raise the error. We'll still need the isabs check on 1494 because it's come through a different code path.

**Review by zooba — COMMENTED** [review]
Two minor changes and I think we're good

**Review by zooba — APPROVED** [review]

**Comment by arhadthedev** [comment]
@zooba I re2orded the docs to reflect your review; would you take a look please

File: Lib/subprocess.py
Comment (eryksun) [inline Lib/subprocess.py]: `CreateProcessW()` will fail with `FileNotFoundError` if the shell path doesn't exist, and it also checks for execute access, which we don't support checking in Python. We do need to check for a relative path. 

```suggestion
                        if not os.path.isabs(comspec):
                            raise FileNotFoundError(
                'shell not found: neither %ComSpec% nor %SystemRoot% is set')
```

Note in terms of security that `os.path.isabs()` is flawed on Windows. It's true for a relative rooted path such as "\\Windows", which depends on the drive or UNC share of the current working directory. However, it's unlikely that `SystemRoot` is set to a relative path. Generally it's either correct or missing.

File: Lib/subprocess.py
Comment (zooba) [inline Lib/subprocess.py]:
```suggestion
                            raise FileNotFoundError('shell not found: neither %ComSpec% nor %SystemRoot% is set')
```

This is better all on one line. (I know PEP 8 says to split it, but it's okay to ignore it this time - there's no important code in or after the message, and it's more readable this way)

Comment (arhadthedev) [reply Lib/subprocess.py]: @eryksun

Comment (eryksun) [reply Lib/subprocess.py]: That's fine by me. I was trying to comply with the 79 character limit in PEP 8 while at least keeping the message itself all on one line.

**Comment by eryksun** [comment]
> Two minor changes and I think we're good

You had suggested that we should try to read the default "ComSpec" value from the system registry hive instead of assuming "%SystemRoot%\\System32\cmd.exe". I thought that could be useful, in case an administrator configured an alternate shell that supports running a `/c` command. But I can also see how reading the registry here is overkill in response to a messed up environment that's missing `%ComSpec%`. You're giving up on this idea, right?

**Comment by zooba** [comment]
> I can also see how reading the registry here is overkill in response to a messed up environment that's missing `%ComSpec%`.

Yeah, I don't really think it's our job to correct for that kind of mess. If we had an existing API to create a fresh environment block as you suggested, then maybe, but it's overkill.

**Comment by zooba** [comment]
Bah, just noticed in the backports that now there's a case where `comspec` may be used without being defined. We should fix and backport that immediately

**Comment by zooba** [comment]
Turns out I can commit directly to the backports, so those will be fine. But we need an `else: comspec = executable` branch in main

**Comment by zooba** [comment]
GH-101712 is the comspec=executable fix for main.

File: Doc/library/subprocess.rst
Comment (gpshead) [inline Doc/library/subprocess.rst:114]: This version number is wrong. 3.11.2 was released before this change was merged. As this PR is in the `main` branch, listing it as 3.12 in these docs is more appropriate. Documentation readers should not be expected to know which patch release versions came before a major release.

We lack the ability to list all versions in versionchanged markers in our docs so have done it this way elsewhere as well when a visible API change has been made in patch releases.

**Comment by arhadthedev** [comment]
> Bah, just noticed in the backports that now there's a case where comspec may be used without being defined. We should fix and backport that immediately

 > This version number is wrong. 3.11.2 was released before this change was merged. As this PR is in the main branch, listing it as 3.12 in these docs is more appropriate. Documentation readers should not be expected to know which patch release versions came before a major release.

Got working on it; the PR will follow in a few hours.

**Comment by arhadthedev** [comment]
> Bah, just noticed in the backports that now there's a case where comspec may be used without being defined. We should fix and backport that immediately

To be fixed by gh-101712. Thank you, @zooba!

> This version number is wrong. 3.11.2 was released before this change was merged. As this PR is in the main branch, listing it as 3.12 in these docs is more appropriate. Documentation readers should not be expected to know which patch release versions came before a major release.

Addressed in gh-101728.

**Pull request merged**
